stdlib/LinearAlgebra/src/adjtrans.jl

```diff
@@ -203,8 +175,8 @@ axes(v::AdjOrTransAbsVec) = (Base.OneTo(1), axes(v.parent)...)
 axes(A::AdjOrTransAbsMat) = reverse(axes(A.parent))
 IndexStyle(::Type{<:AdjOrTransAbsVec}) = IndexLinear()
 IndexStyle(::Type{<:AdjOrTransAbsMat}) = IndexCartesian()
-@propagate_inbounds getindex(v::AdjOrTransAbsVec, i::Int) = wrapperop(v)(v.parent[i-1+first(axes(v.parent)[1])])
-@propagate_inbounds getindex(A::AdjOrTransAbsMat, i::Int, j::Int) = wrapperop(A)(A.parent[j, i])
+@propagate_inbounds getindex(v::AdjOrTransAbsVec{T}, i::Int) where {T} = wrapperop(v)(v.parent[i-1+first(axes(v.parent)[1])])::T
+@propagate_inbounds getindex(A::AdjOrTransAbsMat{T}, i::Int, j::Int) where {T} = wrapperop(A)(A.parent[j, i])::T
 @propagate_inbounds setindex!(v::AdjOrTransAbsVec, x, i::Int) = (setindex!(v.parent, wrapperop(v)(x), i-1+first(axes(v.parent)[1])); v)
 @propagate_inbounds setindex!(A::AdjOrTransAbsMat, x, i::Int, j::Int) = (setindex!(A.parent, wrapperop(A)(x), j, i); A)
```

Comment:
```suggestion
@propagate_inbounds setindex!(v::AdjOrTransAbsVec{T}, x, i::Int) where {T} = (setindex!(v.parent, wrapperop(v)(convert(T, x)), i-1+first(axes(v.parent)[1])); v)
@propagate_inbounds setindex!(A::AdjOrTransAbsMat, x, i::Int, j::Int) = (setindex!(A.parent, wrapperop(A)(convert(T, x)), j, i); A)
```

Though, on reflection, it seems like we built the dispatch tree wrong, and should have instead expected the types to match upfront:
```julia
@propagate_inbounds setindex!(v::AbstractArray{T}, x, i::Int...) where {T} = (setindex!(v, convert(T, x)::T, i...)
@propagate_inbounds setindex!(v::Array{T}, x::T, i::Int...) where {T} = arrayset(v, T, i...)
@propagate_inbounds setindex!(A::AdjOrTransAbsMat{T}, x::T, i::Int...) = (setindex!(A.parent, wrapperop(A)(x), i...); A)
```
Though I suppose that ship has sailed?

Reply:
Doesn't the conversion happen in the called `setindex!` method?

By matching types you mean something like `Adjoint{T,AbstractArray{T}}`? If so, then the issue is that we don't know upfront where the eltype is located within the second type parameter. So it's impossible to restrict the wrapped type to something that thas `T` as `eltype`.

Reply:
The "problem" is that `T` reflects the `eltype` of the adjoint object, not necessarily the eltype of the wrapped object, so your proposal seems "wrong". Think about matrices of matrices, then the eltype of the adjoint is `adjoint{Matrix}`, whereas the eltype of the original is just `Matrix`. I'll see if we have specific tests for such cases.

Reply:
We could ask subtyping for that parameter, but we actually specifically don't want that (since the parent will do it already, as you mentioned). Instead, we can convert the input before the attempted adjoint, so that (hopefully) `adjoint(convert(eltype(S), adjoint(convert(T, x))))` isa `T`

But this can be done later, if needed

Reply:
I'm currently testing the following modification:
```julia
@propagate_inbounds setindex!(v::AdjOrTrans{T,<:AbstractVector{S}}, x, i::Int) =
    (setindex!(v.parent, wrapperop(v)(convert(S, wrapperop(v)(convert(T, x)))), i-1+first(axes(v.parent)[1])); v)
@propagate_inbounds setindex!(A::AdjOrTrans{T,<:AbstractMatrix{S}}, x, i::Int, j::Int) =
    (setindex!(A.parent, wrapperop(A)(convert(S, wrapperop(A)(convert(T, x)))), j, i); A)
```
Is this literally what you have in mind? If it passes locally, should I then push it to this PR? This is going beyond the original PR.

Reply:
Hm, no, it causes trouble. Stuff like
```julia
MethodError: Cannot `convert` an object of type Matrix{Float64} to an object of type LinearAlgebra.Adjoint{Float64, Matrix{Float64}}
```
I think we should go with the PR as is and take another look at `setindex!` later.

Reply:
The `convert(S, _)` is already handled by the parent. But no, I don't think it needs to be part of this PR.

Reply:
> The `convert(S, _)` is already handled by the parent. But no, I don't think it needs to be part of this PR.

True. And I also messed up the brackets a few times. After fixing it several times I realize it's your suggestion above, so I think we could commit that as well.

Reply:
Finally... the suggestion causes issues like the following
```julia
MethodError: Cannot `convert` an object of type Matrix{Float64} to an object of type LinearAlgebra.Adjoint{Float64, Matrix{Float64}}
```
in the tests, so that takes a bit more effort.